Panic in `flow.Task` is not recovered

<!-- Please ensure that you do not include company internal information. -->

**How to categorize this issue?**
<!--
Please select area, kind, and priority for this issue. This helps the community categorizing it.
Replace below TODOs or exchange the existing identifiers with those that fit best in your opinion.
If multiple identifiers make sense you can also state the commands multiple times, e.g.
  /area control-plane
  /area auto-scaling
  ...

"/area" identifiers:     audit-logging|auto-scaling|backup|certification|control-plane-migration|control-plane|cost|delivery|dev-productivity|disaster-recovery|documentation|high-availability|logging|metering|monitoring|networking|open-source|ops-productivity|os|performance|quality|robustness|scalability|security|storage|testing|usability|user-management
"/kind" identifiers:     api-change|bug|cleanup|discussion|enhancement|epic|impediment|poc|post-mortem|question|regression|task|technical-debt|test
-->
/area robustness
/kind bug

**What happened**:

If one `Task` in the [`flow`](https://github.com/gardener/gardener/tree/master/pkg/utils/flow) library panics the complete program panics, as the `Tasks` are run in go routines.

Example: The provider extensions use the `flow` library in the infrastructure controller. In case one of the flow tasks panics the complete `gardener-extension-provider-*` panics and influences other cluster reconciles on the same seed. 

**What you expected to happen**:

The panic should be recoverd (somewhere [here](https://github.com/gardener/gardener/blob/09d35a30058de3b7d7962d857e5c492f71b5c086/pkg/utils/flow/flow.go#L231)), to limit the issue to a flow/reconcile.

This could be done per default or optionally. WDYT?

**How to reproduce it (as minimally and precisely as possible)**:

```go
package main

import (
	"context"
	"fmt"

	"github.com/gardener/gardener/pkg/utils/flow"
)

func main() {
	defer func() {
		// is not used in the task as the task runs in a go routine.
		if r := recover(); r != nil {
			fmt.Println("Recovered in f", r)
		}
	}()

	g := flow.NewGraph("test")
	task := flow.Task{
		Name: "panic test",
		Fn: func(_ context.Context) error {
			panic("hello world")
			return nil
		},
	}

	g.Add(task)
	_ = g.Compile().Run(context.TODO(), flow.Opts{})
}
```
```bash
❯ go run main.go
panic: hello world

goroutine 18 [running]:
main.main.func2({0xae6d70?, 0x37457e3ba2c0?})
        /tmp/flow/main.go:22 +0x25
github.com/gardener/gardener/pkg/utils/flow.(*execution).runNode.func2()
        /go/pkg/mod/github.com/gardener/gardener@v1.139.0/pkg/utils/flow/flow.go:234 +0x184
created by github.com/gardener/gardener/pkg/utils/flow.(*execution).runNode in goroutine 1
       /go/pkg/mod/github.com/gardener/gardener@v1.139.0/pkg/utils/flow/flow.go:231 +0x6a9
exit status 2
```


**Anything else we need to know?**:

**Environment**:

- Gardener version: v1.139.0
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

--- a/pkg/utils/flow/flow.go
+++ b/pkg/utils/flow/flow.go
@@ -229,6 +229,14 @@ func (e *execution) runNode(ctx context.Context, id TaskID) {
 	e.stats.Running.Insert(id)
 
 	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				err := fmt.Errorf("task %q recovered panic: %v", id, r)
+				log.Error(err, "Error")
+				e.done <- &nodeResult{TaskID: id, Error: err}
+			}
+		}()
+
 		start := e.flow.clock.Now().UTC()
 		log.V(1).Info("Started")
 		err := node.fn(ctx)

--- a/pkg/utils/flow/flow_test.go
+++ b/pkg/utils/flow/flow_test.go
@@ -7,6 +7,7 @@ package flow_test
 import (
 	"context"
 	"errors"
+	"fmt"
 	"sync"
 
 	. "github.com/onsi/ginkgo/v2"
@@ -112,6 +113,21 @@ var _ = Describe("Flow", func() {
 			Expect(err).To(HaveOccurred())
 			causes := flow.Causes(err)
 			Expect(causes.Errors).To(ConsistOf(err1, err2))
+		})
+
+		It("should recover from a panic and yield the correct errors", func() {
+			var (
+				panicErr = errors.New("panic error")
+
+				g      = flow.NewGraph("foo")
+				taskId = g.Add(flow.Task{Name: "x", Fn: func(_ context.Context) error { panic(panicErr) }})
+				f      = g.Compile()
+			)
+
+			err := f.Run(ctx, flow.Opts{})
+			Expect(err).To(HaveOccurred())
+			causes := flow.Causes(err)
+			Expect(causes.Errors).To(ConsistOf(fmt.Errorf("task %q recovered panic: %v", taskId, panicErr)))
 		})
 
 		It("should not process any function due to a canceled context", func() {

